### ISSUE TITLE
Inconsistent results on every run with GPU

### ISSUE THREAD

**Issue by Ending2015a** [keras-team/keras#15643]
Repost from tensorflow/tensorflow#52680

**System information**
- OS Platform and Distribution: Colab
- TensorFlow version: 2.6.0 GPU

**Describe the current behavior**
Hello, I'm using TensorFlow 2.4.0 with GPU support on my local machine to train a small CNN model. I can get the same training results when I train it on CPU mode if I fixed the random seed, but always get inconsistent results on GPU mode.

So I test it on Colab, which has TensorFlow 2.6.0 installed in default, but still got inconsistent results on GPU mode.
> edit: Same for TensorFlow 2.7.

**Describe the expected behavior**
The results should be consistent.

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no): no
- Briefly describe your candidate solution(if contributing): x

**Standalone code to reproduce the issue**
Here is the test code
[Colab](https://colab.research.google.com/drive/1Xr_p9rsKM0bk3DDt9CL3L-dN7Vvz179Q?usp=sharing)
Run it for many times with GPU enabled, you will get different results on each run.

**Other info / logs*
I found that only when I use `relu`, `leaky_relu` or `tf.maximum` as the activation function the results change on every run. But if I use `sigmoid`, `tanh`, or other non-relu-like activation functions the results will be the same.

**Comment by sachinprasadhs** [comment]
I could notice the change in final loss on each run, attaching the [gist](https://colab.research.google.com/gist/sachinprasadhs/5751a9c525624aa9132c3975d9dfba05/tensorflow-gradient-bug.ipynb) for reference. Thanks!

**Comment by chenmoneygithub** [comment]
Thanks for reporting this issue! This is actually expected: if model is run on gpu, then cudnn imports this randomness. If you want to get deterministic results, you may check out this api: https://www.tensorflow.org/api_docs/python/tf/config/experimental/enable_op_determinism

But getting a deterministic result is at the cost of performance.

**Comment by Ending2015a** [comment]
@chenmoneygithub 
Thank you for the reply.
So the `relu` is not a deterministic op in gpu mode? because I get the random results only when I use `relu` or other `relu`-like ops, e.g. `leaky_relu`, `tf.maximum`. Besides, I want to test this api on colab, but it seems it only available in tf-nightly, and maybe available after tensorflow 2.8 is released. Is there any api similar to this before tensorflow 2.8? I have already tested setting `os.environ['TF_CUDNN_DETERMINISTIC']='1'` but doesn't work.

**Comment by jvishnuvardhan** [comment]
Adding here a note from the page.

> Note: This API is new and only available via pip install tf-nightly.

So, if you want to test this, you need to install `tf-nightly`. For some reasons, some times colab has difficulty in accessing GPU. So, try testing on your local GPU or cloud GPU. Thanks

**Comment by Ending2015a** [comment]
@jvishnuvardhan 
Hey, thank you for noting me that. How about the version before `tf-nightly` (e.g. tf 2.4~2.7) . Can I have deterministic results on these versions on gpu?

**Comment by jvishnuvardhan** [comment]
@Ending2015a https://www.tensorflow.org/api_docs/python/tf/config/experimental/enable_op_determinism This API is new and only available in `tf-nightly` and not available before (e.g. tf 2.4~2.7).

**Comment by Ending2015a** [comment]
@jvishnuvardhan 
Yes I know it is only available in `tf-nightly`. What I am looking for is the way other than using this api to get the deterministic results on the previous versions (tf 2.4~2.7).

**Comment by chenmoneygithub** [comment]
I do not think we have a similar api in the stable version. 

One question I am curious about is - does the small value fluctuation affect your work? We want to learn more about how severe the problem is, thanks!

**Comment by Ending2015a** [comment]
Hello, @chenmoneygithub
You think that it's just a small fluctuation is because this example code uses a small network which only has 3 small conv layers and only trains for 1000 gradient steps. As I observed in this tiny scripts, each gradient update will cause 1e-6 scale of noises, resulting in 1e-3 scale of fluctuations. While in the real world application we use larger and deeper networks and needs much more training steps. So if we run more training steps, the compounding error becomes huge, and finally we have a completely different models with the same random seeds.

This problem is very severe, especially for researchers who want to reproduce the experiments results or want to debug algorithms. Since each time we get the completely different results, the bugs may not reproducible and we need to spend more time to find the cause. For software developers, it's hard to do CI and unittests since the networks output different results on each run even the random seeds are fixed.

**Comment by Ending2015a** [comment]
So I think TensorFlow/Keras teams can not solve this issue. I decide to switch to PyTorch. Thank you.
